test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,327 @@
+)
+
+type vmInfo struct {
+	nodeName string
+	osType   string
+	ifName   string
+	ip       string
+	ifIndex  string // Used only for Windows
+}
+
+// TestVMAgent is the top-level test which can contain some subtests for
+// VMAgent so they can share setup, teardown.
+func TestVMAgent(t *testing.T) {
+	skipIfFeatureDisabled(t, features.ExternalNode, false, true)
+	skipIfNoVMs(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	vmList, err := setupVMAgentTest(t, data)
+	if err != nil {
+		t.Fatalf("Error when setting up VMAgent test: %v", err)
+	}
+	defer teardownVMAgentTest(t, data, vmList)
+	t.Run("testExternalNode", func(t *testing.T) { testExternalNode(t, data, vmList) })
+}
+
+// setupVMAgentTest creates ExternalNode, starts antrea-agent
+// and returns a list of VMs upon success
+func setupVMAgentTest(t *testing.T, data *TestData) ([]vmInfo, error) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.windowsVMs, testOptions.linuxVMs)
+	nsReturned, _ := data.clientset.CoreV1().Namespaces().Get(context.TODO(), nameSpace, metav1.GetOptions{})
+	saReturned, _ := data.clientset.CoreV1().ServiceAccounts(nameSpace).Get(context.TODO(), serviceAccount, metav1.GetOptions{})
+	t.Logf("Using service account %s", saReturned.Name)
+	t.Logf("Using namespace %s", nsReturned.Name)
+	var vmList []vmInfo
+	if testOptions.linuxVMs != "" {
+		vms := strings.Split(testOptions.linuxVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
+			tempVm := getVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	if testOptions.windowsVMs != "" {
+		vms := strings.Split(testOptions.windowsVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Windows VM: %s", vm)
+			tempVm := getWindowsVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	t.Logf("TestVMAgent setup")
+	for _, vm := range vmList {
+		stopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalNode for VM: %s", vm.nodeName)
+		_, err := createExternalNodeCRD(data, vm.nodeName, vm.ifName, vm.ip)
+		require.NoError(t, err, "Failed to create ExternalNode")
+		startAntreaAgent(t, data, vm)
+	}
+	return vmList, nil
+}
+
+// teardownVMAgentTest deletes ExternalNode, verifies ExternalEntity is deleted
+// and verifies uplink configuration is restored.
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+			var tempVm vmInfo
+			if vm.osType == "Linux" {
+				tempVm = getVMInfo(t, data, vm.nodeName)
+			} else {
+				tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if !assert.Equal(t, vm.ifName, tempVm.ifName, "Unexpected uplink interface name") {
+				return false, nil
+			}
+			if !assert.Equal(t, vm.ip, tempVm.ip, "Unexpected uplink interface IP") {
+				return false, nil
+			}
```

Comment:
`!assert.Equal` will mark the test as failure, is it expected? It looks contradictory to mark failure but keep retrying.

Reply:
We are not running the test with `failfast` mode. So it will retry.
If this marks the test as Failure, then we don't need the assert. 
I will change it to back to comparing two strings.